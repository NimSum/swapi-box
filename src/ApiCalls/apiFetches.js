export const fetchRandomMovie = filmId =>
   fetch(`https://swapi.co/api/films/${filmId}`)
    .then(response => {
      if (!response.ok) {
        throw Error('Error fetching movies')
      } else {
        return response.json();
      }
    })
    