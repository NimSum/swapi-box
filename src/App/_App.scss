* {
  box-sizing: border-box;
}

body {
  margin: 0;
  padding: 0;
}

.App {
  text-align: center;
  height: 100vh;
  background: url('./images/stars-bg.jpg');
}

@keyframes crawl {
  0% {
    top: 0;
    transform: rotateX(20deg)  translateZ(0);
  }
  100% {
    top: -6000px;
    transform: rotateX(25deg) translateZ(-2500px);
  }
}
