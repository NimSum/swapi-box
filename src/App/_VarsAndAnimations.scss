@import url('https://fonts.googleapis.com/css?family=Changa|Roboto');
$content-font: 'Roboto', sans-serif;
$title-font: 'Changa', sans-serif;

$active-btn: #eeeeee;
$button-color-primary: hsl(177, 60%, 66%);
$button-color-secondary: hsl(0, 31%, 57%);
$btn-active-fav: #f08a5d;


$container-bgc-primary: hsl(115, 34%, 81%);
$container-bgc-secondary: hsl(115, 34%, 81%);

$star-wars-font-color: hsl(52, 97%, 58%);
$font-light: #f08a5d;
$font-dark:#222831;
$font-card-title-color: hsl(52, 97%, 58%);
$font-card-desc-color: hsla(183, 100%, 35%, .8);

$card-title-border-color: hsla(0, 0%, 93%, .5);
$card-shadow: 0px 0px 20px hsla(0, 0%, 100%, .3);
$card-bg-opacity: hsla(343, 51%, 48%, .7);

@keyframes crawl {
  0% {
    top: 0;
    transform: rotateX(20deg)  translateZ(0);
  }
  100% {
    top: -6000px;
    transform: rotateX(25deg) translateZ(-2500px);
  }
}
