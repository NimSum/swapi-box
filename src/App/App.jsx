import React, { Component } from 'react';
import { CardContainer } from '../SectionContainers/CardContainer';
import { Header } from '../SectionContainers/Header';
import { CategoryBtnSection } from '../SectionContainers/CategoryBtnSection';
import { fetchRandomMovie, fetchCategory, fetchItem } from '../ApiCalls/apiFetches';
const uuidv4 = require('uuid/v4');

class App extends Component {
  constructor() {
    super();
    this.state = {
      selectedMovie: {},
      loading: false,
      categorySelected: 'home',
      renderCards: [],
      favoriteCount: 0,
      currentPage: 1
    }
  }

  componentDidMount() {
    const randomMovieId = Math.floor(Math.random() * 7) + 1;
    this.setState({ loading: true })
    this.updateFavoriteCount();
    fetchRandomMovie(randomMovieId)
    .then(film => this.setState({ selectedMovie: film, loading: false }))
    .catch(error => console.log(error) )
  }

  updateFavoriteCount = () => {
    const favorites = JSON.parse(localStorage.getItem('favorites')) || [];
    this.setState({ favoriteCount: favorites.length })
  }

  renderFavorites = () => {
    const renderCards = JSON.parse(localStorage.getItem('favorites')) || [];
    renderCards.length && this.setState({ renderCards })
  }

  changeCategory = category => {
    this.setState({ categorySelected: category, renderCards: [] });
    category === 'people' && this.fetchPeople();
    category === 'planets' && this.fetchPlanets();
    category === 'vehicles' && this.fetchVehicles();
    category === 'favorites' && this.renderFavorites();
  }

  fetchPeople = () => 
    fetchCategory('people')
      .then(people => 
        people.map(person => {
          const homeworldInfo = fetchItem(person.homeworld)
            .then(world => ({ worldName: world.name, worldPopulation: world.population }));
          const speciesInfo = fetchItem(...person.species)
            .then(species => ({ speciesName: species.name }));
          return Promise.all([ homeworldInfo, speciesInfo ])
            .then(characterInfo => {
              const card = { ...characterInfo[0], ...characterInfo[1] , ...person, id: uuidv4(), type: 'character' }
              const renderCards = [ ...this.state.renderCards, card ];
              this.setState({ renderCards })
              return renderCards;
            });
        })
      ).catch(error => console.log(error));

  fetchPlanets = () => 
    fetchCategory('planets')
      .then(planets => planets.map(planet => {
        const names = planet.residents.map(resident => 
          fetchItem(resident)
            .then(person => ({ name: person.name, id: uuidv4() })));
        return Promise.all(names)
          .then(names => {
            const card = ({ residentNames: names, ...planet, id: uuidv4(), type: 'planet', imgNum: Math.floor(Math.random() * 7) + 1 })
            const renderCards = [...this.state.renderCards, card]
            this.setState({ renderCards })
            return renderCards;
          })
      })).catch(error => console.log(error));

  fetchVehicles = () => 
    fetchCategory('vehicles')
      .then(vehicles => vehicles.map(vehicle => {
        const vehicleCard = { ...vehicle, id: uuidv4(), type: 'vehicle' }
        const renderCards = [...this.state.renderCards, vehicleCard ]
        this.setState({ renderCards })
        return renderCards;
      })).catch(error => console.log(error));

  fetchMore = () => {
    // console.log()
  }

  render() {
    return (
      <div className="App">
        < Header category={ this.state.categorySelected }/>
        < CardContainer 
          loading= { this.state.loading }
          movie = { this.state.selectedMovie }
          cards={ this.state.renderCards }
          updateFavoriteCount={ this.updateFavoriteCount } 
          category={ this.state.categorySelected } />
        < CategoryBtnSection 
          currCategory={ this.state.categorySelected }
          changeCategory={ this.changeCategory }
          favoriteCount={ this.state.favoriteCount }
          updateFavoriteCount={ this.props.updateFavoriteCount } />
      </div>
    )
  }
}

export default App;
