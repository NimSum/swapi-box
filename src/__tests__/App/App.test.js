import React from 'react';
import { shallow } from 'enzyme';
import App from '../../App/App';
import * as fetchMethods from '../../ApiCalls/apiFetches';

describe('App component', () => {

  let wrapper;
  let spyFavCount;
  let fetchPlanetsSpy;
  let fetchVehiclesSpy;
  let renderFavoritesSpy;
  let fetchPeopleSpy;
  let setStateSpy;

  let categoryFetchSpy =  jest.spyOn(fetchMethods, 'fetchCategory');
  let fetchMovieSpy = jest.spyOn(fetchMethods, 'fetchRandomMovie');

  beforeEach(() => {
    wrapper = shallow(< App />)
    spyFavCount = jest.spyOn(wrapper.instance() , 'updateFavoriteCount');
    fetchPlanetsSpy = jest.spyOn(wrapper.instance(), 'fetchPlanets');
    fetchVehiclesSpy = jest.spyOn(wrapper.instance(), 'fetchVehicles');
    renderFavoritesSpy = jest.spyOn(wrapper.instance(), 'renderFavorites')
    fetchPeopleSpy = jest.spyOn(wrapper.instance(), 'fetchPeople');
    setStateSpy = jest.spyOn(wrapper.instance(), 'setState');
  });

  afterEach(() => {
    categoryFetchSpy.mockClear();
    fetchMovieSpy.mockClear();
    setStateSpy.mockClear();
  })

  it('Should have default states', () => {
    const mockState = {
      selectedMovie: {},
      loading: false,
      categorySelected: '',
      renderCards: [],
      favoriteCount: 0
    }
    expect(wrapper.state()).toEqual(mockState);
  });

  it('Should match snapshot taken', () => {
    expect(wrapper).toMatchSnapshot();
  })
  
  it('Should invoke methods when mounted', () => {
    expect(fetchMovieSpy).toHaveBeenCalledTimes(1);
    wrapper.instance().componentDidMount();
    expect(spyFavCount).toHaveBeenCalledTimes(1);
  });

  it('Should update favorite count', () => {
    localStorage.setItem('favorites', JSON.stringify([{ item: 1}, { item: 2}]))
    wrapper.instance().updateFavoriteCount();
    expect(wrapper.state().favoriteCount).toEqual(2);
  });

  describe('Setting Category', () => {
    it('Should trigger fetch people cards and change category to people', () => {
      wrapper.instance().changeCategory('people');
      expect(wrapper.state().categorySelected).toEqual('people');
      expect(fetchPeopleSpy).toHaveBeenCalledTimes(1);
    });
  
    it('Should trigger fetch vehicle cards and change category to vehicles', () => {
      wrapper.instance().changeCategory('vehicles');
      expect(wrapper.state().categorySelected).toEqual('vehicles');
      expect(fetchVehiclesSpy).toHaveBeenCalledTimes(1);
    });
  
    it('Should trigger fetch planet cards and change category to planets', () => {
      wrapper.instance().changeCategory('planets');
      expect(wrapper.state().categorySelected).toEqual('planets');
      expect(fetchPlanetsSpy).toHaveBeenCalledTimes(1);
    });
  
    it('Should trigger render favorited cards and change category to favorites', () => {
      wrapper.instance().changeCategory('favorites');
      expect(wrapper.state().categorySelected).toEqual('favorites');
      expect(renderFavoritesSpy).toHaveBeenCalledTimes(1);
    });
  })

  describe('Fetch methods', () => {

    const localFetch = response =>
      window.fetch = jest.fn().mockImplementation(() => 
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(response)
      }));

    it('Should fetch people', () => {
      wrapper.instance().fetchPeople()
       .then(result => expect(setStateSpy).toHaveBeenCalledTimes(1));
      expect(categoryFetchSpy).toHaveBeenCalledWith('people');
    });
  
    it('Should fetch planets', () => {
      wrapper.instance().fetchPlanets()
        .then(result => expect(setStateSpy).toHaveBeenCalledTimes(1)); 
      expect(categoryFetchSpy).toHaveBeenCalledWith('planets');
    });

    it('Should fetch vehicles', () => {
      const mockResponse = { results: [{ name: 'blah'}]};
      window.fetch = localFetch(mockResponse);
      wrapper.instance().fetchVehicles()
        .then(result => expect(setStateSpy).toHaveBeenCalledTimes(1));
      expect(categoryFetchSpy).toHaveBeenCalledWith('vehicles');
    });

    it('Should setstate render favorites', () => {
      const mockCards = [{ name: "Obi-wan Kenobi" }, { name: "Darth Vader" }]
      localStorage.setItem('favorites', JSON.stringify(mockCards));
      wrapper.instance().renderFavorites();
      expect(setStateSpy).toHaveBeenCalledTimes(2);
      expect(wrapper.state().renderCards).toEqual(mockCards);
    });
  });
})




