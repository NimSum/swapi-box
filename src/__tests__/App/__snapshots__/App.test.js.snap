// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should match snapshot taken 1`] = `
<div
  className="App"
>
  <Header
    category="home"
  />
  <CardContainer
    cards={Array []}
    category="home"
    loading={true}
    movie={Object {}}
    updateFavoriteCount={[Function]}
  />
  <CategoryBtnSection
    changeCategory={[Function]}
    currCategory="home"
    favoriteCount={0}
  />
</div>
`;
